### PR TITLE
AWS migration overview

### PR DESCRIPTION
**Pull request by nickmarden**

### CODE REVIEW

File: _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+title: Rehosting a Heavily-Used SaaS Search Engine to AWS
+layout: blog_post
+date: 2016-07-23
+description: How we moved search.usa.gov from a traditional datacenter to Amazon Web Services (AWS), and how that move saved money while improving operational quality
+author: Nick Marden
+---
+
+## Introduction
+
+[search.usa.gov](http://search.digitalgov.gov) is a free service provided by the United States' [General Services Administration](http://www.gsa.gov/) to more than 2,600 other federal, state, or local government agencies. This service allows agencies to easily and intuitively configure a search engine results page (SERP) experience that covers their domains, their documents, and their social media footprint, so that citizens can easily search government web sites to find the information they need. Customers include the [Internal Revenue Service](http://find.irs.gov/search?utf8=%E2%9C%93&affiliate=irs&query=tax+extension&commit=Search), the [Department of Defense](http://search.defense.gov/search?utf8=%E2%9C%93&affiliate=dod-search&query=f-35+fighter+jet), and the [White House](https://search.whitehouse.gov/search?query=birth+certificate&op=Search&affiliate=wh) among others.
```
Comment (noremmie) [inline _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown]: LOL - birth certificate

File: _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+The final savings calculation looked like this:
+
+* Old CDN/WAF provider: $16,000 per month
+* CloudFront distribution and OWASP-enabled proxy servers: $400 per month (including traffic costs)
+
+That's a **97.5%** savings, we were delighted to return to the operating budget of the project for more important things like developing new features!
```
Comment (noremmie) [inline _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown]: I think you don't want the comma after "savings".

**Comment by noremmie** [comment]
This looks great to me - I only really had one grammar suggestion which I put in its own comment.

I would classify the OWASP stuff we're using as "rules" rather than "software", and I would say that modsecurity is the software. But that might be making it too complicated since the OWASP rules are the important part. I'm guessing you're not making the distinction because that would make the post too complicated. I agree that we might be able to make a post out of that work in particular, so I think it's fine to refer to "OWASP software" here.

I wouldn't have thought to mention the ATO process, but I'm glad you did. I'm guessing the message is "yes, Government Agencies who might want to hire us in the future... we know about your compliance processes."

**Comment by nickmarden** [comment]
Thanks @noremmie - I took your small edit into account and also applied some more varnish of my own. In particular the AWS network diagram is now a bit more complete, and some wordier passages have been unmuddled a bit. If you're able to look at it one more (and, I hope, final) time, I'd appreciate it.

File: _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+In this post we'll explain how we re-architected and migrated the search.usa.gov infrastructure in Amazon Web Services (AWS) in order to:
+
+  * Reduce datacenter costs by 40% per month
+  * Remove an expensive Content Delivery Network (CDN)/Web Application Firewall (WAF) provider from our framework, reducing monthly CDN costs by a whopping **97.5%**
+  * Improve reliability by creating self-healing servers and distributing the service across four physically isolated datacenters
+
+## Original Architecture
+
+In the prior search.usa.gov datacenters - one in Chicago and one in Virgina - we had a pools of high-powered Dell "pizza box" servers running a mish-mosh of services in a composition that had been tuned to the observed traffic patterns of search.usa.gov:
```
Comment (noremmie) [inline _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown]: I had to look this up, but I don't think "mish-mosh" should be hyphenated.

Comment (nickmarden) [inline _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown]: :themoreyouknowgif:

File: _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+The final savings calculation looked like this:
+
+* Old CDN/WAF provider: $16,000 per month
+* CloudFront distribution and OWASP-enabled proxy servers: $400 per month (including traffic costs)
+
+That's a **97.5%** savings we were delighted to return to the operating budget of the project for more important things like developing new features!
+
+In other blog posts we discuss in more detail the complexities of supporting [SSL certificates for our government customers's hostnames]({% post_url 2016-07-11-using-lets-encrypt-for-a-multi-tenant-web-application %}) and how we managed to comply with the [DNSSEC requirement]({% post_url 2016-07-09-making-dnssec-work-for-an-aws-hosted-saas %}) for government agencies. Those posts go into much more technical detail and are worth reading if you're interested in how we solve the security challenges of providing SaaS services for thousands of government agencies.
```
Comment (noremmie) [inline _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown]: s/solve/solved/? I admit that I changed verb tenses continuously throughout the SSL post, but "solved" sounds more correct to me here.

File: _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+* Vulnerability testing of individual servers
+* Compliance testing of individual servers
+* Penetration testing of individual servers and exposed pool endpoints
+
+Since individual servers are identified by IP address, these tests get a little bit more complicated when individual servers can be rebuilt without warning due to server failure, load spikes, or data center outages - because the new server will often come online with a different DHCP'd IP address than the one it replaced. Similarly, if a new server is spun up in response to increased load, it will have an IP address that is unknown to the security testing infrastructure and will cause unnecesssary red flags.
+
+We have worked closely with GSA security personnel to understand their security auditing requirements and to help them understand the new terrain in which we are operating. We hope this two-way conversation will lead to updated security practices that can take the dynamic nature of cloud hosting environments into account without sacrificing completeness or quality.
```
Comment (noremmie) [inline _posts/2016-07-23-migrating-a-heavily-used-government-service-to-aws.markdown]: I continue to love this potential outcome even though I never would have thought to mention it in a post!

**Comment by noremmie** [comment]
I combed out a couple more nits, but otherwise this still looks great to me 👍 

Thanks for the hat tip! I'm going to hunker down in preparation for the increased fandom.

**Pull request merged**